File: Python/84. Largest Rectangle in Histogram.py
# 84. Largest Rectangle in Histogram.      HARD

# Topic: Array, Stack, Monotonic Stack.

"""
### Task:
---
Given an array of integers heights representing the histogram's bar height where the width 
of each bar is 1, return the area of the largest rectangle in the histogram.

Example 1:
Input: heights = [2,1,5,6,2,3]
Output: 10
Explanation: The above is a histogram where width of each bar is 1.
The largest rectangle is shown in the red area, which has an area = 10 units.

Example 2:
Input: heights = [2,4]
Output: 4

Constraints:
1 <= heights.length <= 10^5
0 <= heights[i] <= 10^4


### Testcase:
---
[2,1,5,6,2,3]
[2,4]

### Code:
---
class Solution(object):
    def largestRectangleArea(self, heights):
        '''
        :type heights: List[int]
        :rtype: int
        '''
"""
### Solution:-------------------------------------------------------------------

class Solution(object):
    def largestRectangleArea(self, heights):
        '''
        :type heights: List[int]
        :rtype: int
        '''
        # Initialize an empty stack to store indices of bars
        stack = []
        # Initialize max_area to keep track of the largest rectangle area found
        max_area = 0
        # Append a sentinel bar of height 0 to handle remaining bars after the loop
        heights.append(0)

        # Iterate over each bar in the histogram
        for i in range(len(heights)):
            # While stack is not empty and current bar's height is less than the top bar in the stack
            while stack and heights[i] < heights[stack[-1]]:
                # Pop the top bar from the stack and calculate its height
                h = heights[stack.pop()]
                # Calculate the width of the rectangle with the popped bar as height
                w = i if not stack else i - stack[-1] - 1
                # Update max_area if the calculated area is larger
                max_area = max(max_area, h * w)
            # Push the current index onto the stack
            stack.append(i)

        # Return the maximum area found
        return max_area


### Description:
'''
The "Largest Rectangle in Histogram" problem can be efficiently solved using a stack, specifically a monotonic stack. The idea is to maintain a stack of indices of the bars, and calculate the area of rectangles with the stack bars as heights. When a smaller bar is encountered, it triggers the calculation of the area for bars in the stack that are taller than the current one.

Here's a detailed approach:

1. **Initialize Variables**: Create a stack to store indices. Initialize a variable to keep track of the maximum area found.

2. **Iterate Over the Heights**: Go through each bar in the histogram:
   - While the stack is not empty and the current bar's height is less than the height of the bar at the top of the stack, pop from the stack and calculate the area formed by the popped bar. The width of the area is the difference between the current index and the index just below the top of the stack (after popping), minus one.
   - Push the current index onto the stack.

3. **Handle Remaining Bars**: After the loop, some bars may remain in the stack. These bars were never smaller than any bar to their right. For each of these, calculate the area considering the width extends to the end of the histogram.

4. **Return the Maximum Area**.

### Explanation:
- We append a sentinel bar of height 0 to handle any remaining bars in the stack after iterating through the list.
- The `while` loop inside the `for` loop calculates the area for each bar popped from the stack.
- The width calculation changes depending on whether the stack is empty or not after popping. If the stack is empty, the width is the current index `i`; otherwise, it's the distance between the current index and the index below the top of the stack.

This algorithm has a time complexity of O(n), as each bar is pushed and popped from the stack exactly once. The space complexity is also O(n) for the stack in the worst case.

### Breakdown of the Code:

1. **Stack Initialization**: An empty stack is used to store the indices of bars. This stack helps to keep track of bars that might form part of a larger rectangle.

2. **Max Area Initialization**: `max_area` is used to record the largest rectangle area found during the iteration.

3. **Append Sentinel Bar**: Adding a bar of height 0 at the end ensures that any remaining bars in the stack are processed, as they will all be greater than 0.

4. **Iterating Over the Heights**: The loop goes through each bar in the histogram.

5. **Processing the Stack**: If the current bar is lower than the bar at the top of the stack, we pop from the stack, calculate the area with the popped bar as height, and update `max_area` if this area is larger than the current `max_area`.

6. **Pushing to the Stack**: If the current bar is taller than the bar at the top of the stack (or the stack is empty), we push the current index onto the stack.

7. **Returning the Result**: After processing all bars, the largest rectangle area found is returned.

'''

File: Cpp/84. Largest Rectangle in Histogram.cpp
// 84. Largest Rectangle in Histogram.       - HARD -

// Topic: Array, Stack, Monotonic Stack.


/*
### Task:
---
Given an array of integers heights representing the histogram's bar height where the width of each bar is 1, return the
area of the largest rectangle in the histogram.

Example 1:
Input: heights = [2,1,5,6,2,3]
Output: 10
Explanation: The above is a histogram where width of each bar is 1.
The largest rectangle is shown in the red area, which has an area = 10 units.

Example 2:
Input: heights = [2,4]
Output: 4

Constraints:
1 <= heights.length <= 10^5
0 <= heights[i] <= 10^4


### Testcase:
---
[2,1,5,6,2,3]
[2,4]


### Code:
---
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        
    }
};

*/
// Solution: --------------------------------------

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        // Initialize a stack to keep indices of bars
        stack<int> stk;
        // Variable to keep track of the maximum area found
        int maxArea = 0;
        // Get the number of bars in the histogram
        int n = heights.size();

        // Iterate through each bar in the histogram
        for (int i = 0; i <= n; ++i) {
            // While the stack is not empty and the current bar's height is less than
            // the height of the bar at the top of the stack
            while (!stk.empty() && (i == n || heights[stk.top()] > heights[i])) {
                // Get the height of the bar at the top of the stack
                int height = heights[stk.top()];
                // Pop the top element from the stack
                stk.pop();
                // Calculate the width of the rectangle with the popped bar as the shortest bar
                // If the stack is empty, the width is 'i'
                // Otherwise, the width is the difference between 'i' and the new top of the stack - 1
                int width = stk.empty() ? i : i - stk.top() - 1;
                // Update the maximum area found
                maxArea = max(maxArea, height * width);
            }
            // Push the current index onto the stack
            stk.push(i);
        }

        // Return the maximum area found
        return maxArea;
    }
};




// Description: ===================================
/*
To solve the "Largest Rectangle in Histogram" problem, we can use a stack to efficiently track the indices of bars in the histogram. 
The stack helps us determine the boundaries of rectangles with the current bar as the shortest bar. We use a monotonic stack approach, 
which keeps the indices of the bars in the stack in increasing order of their heights.

Here's a step-by-step approach:

1. **Initialize**: Create an empty stack to store indices of bars. Initialize `maxArea` as 0 to keep track of the maximum area found.

2. **Iterate through Each Bar**: For each bar in the histogram, do the following:
   - While the stack is not empty and the current bar's height is less than the height of the bar at the top of the stack, it means we've 
   found a right boundary for the rectangle containing the bar at the top of the stack. Pop the top of the stack, calculate the area of 
   the rectangle using the height of the popped bar, and update `maxArea` if this area is larger.
   - Push the current index onto the stack.

3. **Handle Remaining Bars**: After iterating through all bars, some bars might not have been popped from the stack. These bars don't 
have a right boundary smaller than themselves. For each remaining bar in the stack, calculate the area of the rectangle with the bar as 
the shortest one and update `maxArea` if necessary.

4. **Return Result**: After processing all bars, `maxArea` will contain the area of the largest rectangle.

### Description:

This solution leverages a monotonic stack to find the largest rectangle in a histogram efficiently. By maintaining a stack of indices 
of bars in increasing order of heights, the algorithm can quickly determine the left and right boundaries for the largest rectangle 
that can be formed with each bar as the shortest bar in that rectangle. When a bar is encountered that is shorter than the bar at the 
top of the stack, it signifies a right boundary for rectangles involving bars in the stack, allowing for the calculation of areas and 
updating the maximum area found. This approach ensures that each bar is pushed and popped from the stack exactly once, leading to a time 
complexity of O(n), making it highly efficient for large histograms.

*/
